pytorch_lightning/loops/dataloader/evaluation_loop.py

```diff
@@ -256,10 +256,21 @@ def _on_evaluation_epoch_end(self) -> None:
 
     def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
-        if self._data_fetcher is not None and self._data_fetcher.dataloader_iter is not None:
-            state_dict["dataloader_state_dict"] = asdict(self._data_fetcher.dataloader_iter.previous_state)
+        if self._data_fetcher is None or self._num_val_batches_reached():
+            return state_dict
+        if self._has_completed_evaluation_batch():
+            state = self._data_fetcher.dataloader_iter.state
+        else:
+            state = getattr(self._data_fetcher.dataloader_iter, "previous_state", {})
+        state_dict["dataloader_state_dict"] = asdict(state) if is_dataclass(state) else state
         return state_dict
 
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
         self._dataloader_state_dict = state_dict.get("dataloader_state_dict", {})
+
+    def _has_completed_evaluation_batch(self) -> bool:
+        return self.epoch_loop.batch_progress.current.ready == self.epoch_loop.batch_progress.current.completed
```

Comment:
This method should be in the epoch loop instead

Reply:
c.f conversation above with @awaelchli. It was originally in the epoch loop but requested to be moved up :)